### ISSUE TITLE
How to assist browsers with caching large files

### ISSUE THREAD

**Issue by Gerharddc** [sammhicks/picoserve#45]
I am trying to serve a large single-page application (SPA) squished into a single file from an MCU using picoserve. Since this page is fairly large and very static, I would ideally like browsers to cache it. Since the browser probably won't cache the result for the root path, I have set the root path to simply redirect to a path called "app" with "Cache-Control: public". I have done this as follows:

```
.route("/", get(|| picoserve::response::Redirect::to("/app")))
.route(
    "/app",
    get(|| {
        picoserve::response::File::with_content_type_and_headers(
            "text/html; charset=utf-8",
            include_bytes!("index.html.br"),
            &[("Content-Encoding", "br"), ("Cache-Control", "public")],
        )
    }),
)
```

Unfortunately, when using dev mode on my browser, I can see that it still sends a get request for "/app" every time I load the page. However, it seems I am lucky that the browser includes "If-None-Match" in the header so that the server can choose to respond with 304. But unfortunately it then seems picoserve does not manage exploit that opportunity and instead responds 200 with the full file.

Looking at https://github.com/sammhicks/picoserve/blob/5e462d70b9eab91fe8d8b56973462458f79ead65/src/response/fs.rs#L150, it seems that replying with 304 in this situation should already be supported but for some reason that code never seems to get run. Do I need to set anything more to enable that?

P.S. any other advise on how to best achieve my goal with picoserve?

**Comment by sammhicks** [comment]
Call `get_service` instead of `get`, passing the `File` directly, i.e.

```rust
.route(
    "/app",
        get_service(picoserve::response::File::with_content_type_and_headers(
            "text/html; charset=utf-8",
            include_bytes!("index.html.br"),
            &[("Content-Encoding", "br"), ("Cache-Control", "public")],
        )),
)
```

This does seem to be a footgun. Given how `File` is designed to be used with content generated at compile time, I'm tempted to remove the `IntoResponse` implementation of `File` and provide an example.

**Comment by Gerharddc** [comment]
@sammhicks thanks, that seems to fix the issue with https://github.com/sammhicks/picoserve/blob/5e462d70b9eab91fe8d8b56973462458f79ead65/src/response/fs.rs#L150 never even getting hit. It would indeed be useful to add an example to illustrate the difference between `get` and `get_service` since I had no idea that was a thing.

That being said, I am still having issues with the `|etag| self.etag == etag.as_raw()` test not passing even though it should. I still don't a 304 response.

From the below screenshot it is clear that "If-None-Match" should in fact match the Etag for the page as they are reported as being the same by the browser.

![image](https://github.com/user-attachments/assets/35d934a6-8d35-4ce4-a70e-270572decc82)

**Comment by sammhicks** [comment]
Oops, there's a bug! The equality check for `ETag` doesn't strip the quotes before decoding and comparing the request header, thus always returns `false`.

The fix is in the newly released [`v0.12.0`](https://github.com/sammhicks/picoserve/tree/picoserve-v0.12.0).

Thanks for looking into it! :+1: :smile:

**Comment by Gerharddc** [comment]
@sammhicks that seems to have fixed the problem, thanks!!
